lib/event.c

```diff
@@ -909,9 +1245,20 @@ static int fd_poll(struct event_loop *m, const struct timeval *timer_wait,
 	if (num < 0 && errno == EINTR)
 		*eintr_p = true;
 
+#if defined(USE_EPOLL) &&                                                      \
+	(defined(HAVE_EPOLL_WAIT) || defined(HAVE_EPOLL_PWAIT) ||              \
+	 defined(HAVE_EPOLL_PWAIT2))
+	if (num > 0 && m->awakened) {
```

Comment:
`AWAKEN()` writes this value, ostensibly with `m->mtx` held. But here, we are in `fd_poll()` which is called without holding `m->mtx`. Thus we have a bit of a data race here. Were you aware of this, and if so, what's the rationale behind this code? I assume if you left it like this you determined it was OK to do so; I would appreciate if you could explain why.

The idea behind `AWAKEN()` and the pipe poker fd was to include an fd in the poll set that that can be used by non-thread-local callers to force a return from `poll()`, if we are sleeping in it, in order to handle new work. I am not sure I fully understand what this new `awaken` variable is accomplishing so if you could explain that as well I would appreciate it.

Not saying this is wrong by the way, just want to make sure we are doing the right thing here.

Reply:
This is a good point. Let me explain why the `m->awakened` variable is added first, and then discuss about data race.

## Why the `m->awakened` variable is added
```c
#if EPOLL_ENABLED
	if (num > 0 && m->awakened) {
		/* AWAKENed, exhaust io_pipe */
		while (read(m->io_pipe[0], &trash, sizeof(trash)) > 0)
			;
		m->awakened = false;
	}
#else
	if (num > 0 && m->handler.copy[count].revents != 0 && num--)
		while (read(m->io_pipe[0], &trash, sizeof(trash)) > 0)
			;
#endif
```
When poll() is used, we maintain a pollfd array and append the io_pipe[0] at the end of it. Therefore, by checking `.renvents` of the last pollfd `m->handler.copy[count]`, we can judge whether poll() is returned because of an `AWAKEN()`.

When using epoll APIs, all file descriptors under polling are maintained in an epoll set but not a self-maintained array. When epoll_wait is returned, we can't judge whether it is returned normally or returned from `AWAKEN()` by simply checking the last element of a self-maintained array as we did in poll() case. An naive and straightforrward way to judge is: scan the `revents` array returned by epoll_wait, find the `struct epoll_event` element correspoing to io_pipe[0], and check the `.events` of it. However, this "scanning" way is too cumbersome and contradicts with the idea of removing "scanning" process of polling. Therefore, I added an `awakened` variable to make the judgement clearer and faster.

## About data race on `m->awakened`

In terms of data race, I think code above works fine even without locking in practice. The side effect of reading wrong `m-awakened` value is extra exhausting operation on `io_pipe`, which will not raise any error. Also, contending on `m->awakened` doesn't affect `write` on `io_pipe` to wake a sleeping method. I had executed topotests many times and didnt observe error.

Indeed this is kind of dangerous and probably I was wrong. Nevertheless, I have wrapped all access of `m->awakened` with a new mutex `awakened_mtx`:
```c
/* lib/frrevent.h */
struct event_loop {
	...
#if EPOLL_ENABLED
	bool awakened;
	pthread_mutex_t awakened_mtx;
#endif
}

/* lib/event.c */
#if EPOLL_ENABLED
/* When epoll APIs are used, the AWAKEN macro is supposed to be called with m->awakened mtx held */
#define AWAKEN(m)                                                              \
	do {                                                                   \
		const unsigned char wakebyte = 0x01;                           \
		pthread_mutex_lock(&m->awakened_mtx);                                   \
		m->awakened = true;                                            \
		pthread_mutex_unlock(&m->awakened_mtx);                                 \
		write(m->io_pipe[1], &wakebyte, 1);                            \
	} while (0)
#else
...

/* Init, free, and access on m->awakend_mtx are modified accordingly */
```

Reply:
Please add a synopsis of this analysis as a in place code comment.  I do not want to loose this discussion when someone is looking at the code in 5 years

Reply:
Added comments in frrevent.h about purpose of `awaken` field of `struct event_loop`.